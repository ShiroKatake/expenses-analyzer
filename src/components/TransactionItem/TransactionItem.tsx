import { useEffect, useRef } from "react";
import { stringToDate } from "../../utils/stringToDate";
import { InlineTextInput } from "../InlineTextInput/InlineTextInput";
import { Transaction } from "../../types/transaction";

interface ITransactionItemProps {
  transaction: Transaction;
  index: number;
}

export const TransactionItem = ({ transaction, index }: ITransactionItemProps) => {
  const rowRef = useRef<HTMLTableRowElement>(null);
  let isHovered = false;

  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.repeat || !isHovered) {
        return;
      }
      if (event.code === "KeyT") {
        console.log(`Add tag item ${index}`);
      }
    };

    if (rowRef.current !== null) {
      document.addEventListener("keydown", handleKeyDown);
      rowRef.current.addEventListener("mouseenter", () => (isHovered = true));
      rowRef.current.addEventListener("mouseleave", () => (isHovered = false));
    }

    return function cleanup() {
      if (rowRef.current !== null) {
        document.removeEventListener("keydown", handleKeyDown);
        rowRef.current.removeEventListener("mouseenter", () => (isHovered = true));
        rowRef.current.removeEventListener("mouseleave", () => (isHovered = false));
      }
    };
  }, []);

  return (
    <tr ref={rowRef}>
      <td>{stringToDate(transaction.date)}</td>
      <td>
        <InlineTextInput transactionName={transaction.name} transactionIndex={index} />
      </td>
      <td>{transaction.amount}</td>
      <td>{transaction.balance}</td>
      <td></td>
    </tr>
  );
};
