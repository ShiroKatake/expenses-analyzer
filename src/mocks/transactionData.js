module.exports = [
  {
    date: "01-Apr-20",
    name: "Purchase A",
    amount: 10,
    balance: 110,
  },
  {
    date: "02-Apr-20",
    name: "Purchase B",
    amount: -13.23,
    balance: 96.77,
  },
  {
    date: "02-Apr-20",
    name: "Purchase C",
    amount: -200,
    balance: -103.23,
  },
];
