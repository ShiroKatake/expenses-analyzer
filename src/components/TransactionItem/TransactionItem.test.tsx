import { cleanup, fireEvent, render, screen } from "@testing-library/react";
import { TransactionItem } from "./TransactionItem";
import transactionData from "../../mocks/transactionData";

describe("TransactionItem", () => {
  const component = render(
    <table>
      <tbody>
        {transactionData.map((transaction, index) => {
          return <TransactionItem key={index} transaction={transaction} index={index} />;
        })}
      </tbody>
    </table>
  );
  afterEach(() => {
    cleanup();
  });

  it("should update transaction name after editing input", () => {
    let inputDiv = screen.getByText("Purchase A");
    fireEvent.mouseOver(inputDiv);
    fireEvent.keyPress(inputDiv, { key: "KeyT", charCode: 84 });
    expect(component).toMatchSnapshot();
  });
});
