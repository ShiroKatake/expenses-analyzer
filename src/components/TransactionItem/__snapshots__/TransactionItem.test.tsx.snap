// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionItem should update transaction name after editing input 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table>
        <tbody>
          <tr>
            <td
              data-testid="date"
            >
              01 Apr 2020
            </td>
            <td
              data-testid="name"
            >
              <div
                class="_1EEDX _3sXmF sc-bdvvtL iErRBy"
              >
                Purchase A
              </div>
            </td>
            <td
              data-testid="amount"
            >
              10
            </td>
            <td
              data-testid="balance"
            >
              110
            </td>
            <td />
          </tr>
          <tr>
            <td
              data-testid="date"
            >
              02 Apr 2020
            </td>
            <td
              data-testid="name"
            >
              <div
                class="_1EEDX _3sXmF sc-bdvvtL iErRBy"
              >
                Purchase B
              </div>
            </td>
            <td
              data-testid="amount"
            >
              -13.23
            </td>
            <td
              data-testid="balance"
            >
              96.77
            </td>
            <td />
          </tr>
          <tr>
            <td
              data-testid="date"
            >
              03 Apr 2020
            </td>
            <td
              data-testid="name"
            >
              <div
                class="_1EEDX _3sXmF sc-bdvvtL iErRBy"
              >
                Purchase C
              </div>
            </td>
            <td
              data-testid="amount"
            >
              -200
            </td>
            <td
              data-testid="balance"
            >
              -103.23
            </td>
            <td />
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table>
      <tbody>
        <tr>
          <td
            data-testid="date"
          >
            01 Apr 2020
          </td>
          <td
            data-testid="name"
          >
            <div
              class="_1EEDX _3sXmF sc-bdvvtL iErRBy"
            >
              Purchase A
            </div>
          </td>
          <td
            data-testid="amount"
          >
            10
          </td>
          <td
            data-testid="balance"
          >
            110
          </td>
          <td />
        </tr>
        <tr>
          <td
            data-testid="date"
          >
            02 Apr 2020
          </td>
          <td
            data-testid="name"
          >
            <div
              class="_1EEDX _3sXmF sc-bdvvtL iErRBy"
            >
              Purchase B
            </div>
          </td>
          <td
            data-testid="amount"
          >
            -13.23
          </td>
          <td
            data-testid="balance"
          >
            96.77
          </td>
          <td />
        </tr>
        <tr>
          <td
            data-testid="date"
          >
            03 Apr 2020
          </td>
          <td
            data-testid="name"
          >
            <div
              class="_1EEDX _3sXmF sc-bdvvtL iErRBy"
            >
              Purchase C
            </div>
          </td>
          <td
            data-testid="amount"
          >
            -200
          </td>
          <td
            data-testid="balance"
          >
            -103.23
          </td>
          <td />
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
